### PR TITLE
Deprecating FCM Direct Channel API

### PR DESCRIPTION
**Pull request by chliangGoogle**

### CODE REVIEW

**Comment by paulb777** [comment]
@chliangGoogle Is this ready for review?

**Comment by chliangGoogle** [comment]
We are still waiting for launchcal but the code is ready for review.

File: Firebase/Messaging/Public/FIRMessaging.h [hunk 1]
```diff
@@ -252,7 +252,8 @@ NS_SWIFT_NAME(MessagingDelegate)
 /// Handle data messages received via FCM direct channel (not via APNS).
 - (void)messaging:(FIRMessaging *)messaging
     didReceiveMessage:(FIRMessagingRemoteMessage *)remoteMessage
-NS_SWIFT_NAME(messaging(_:didReceive:));
+NS_SWIFT_NAME(messaging(_:didReceive:))
+__deprecated_msg("Downstream message in direct channel is deprecated, please use APNS channel.");
```
Comment (paulb777) [inline Firebase/Messaging/Public/FIRMessaging.h]: Is there a URL that describes how to transition?

**Comment by ryanwilson** [comment]
It looks like `FIRMessagingMessageStatus` was missed - please deprecate that as well. Also I believe this needs formatting, some of the lines with deprecation messages are quite long.

**Comment by chliangGoogle** [comment]
FIRMessagingMessageStatus is used by FIRMessagingMessageInfo that is used by appDidReceiveMessage: that is not deprecated.
I've updated the API review doc for this change. And run ./script/format on the header file.

File: Firebase/Messaging/Public/FIRMessaging.h (hunk 1 above)
Comment (chliangGoogle) [reply Firebase/Messaging/Public/FIRMessaging.h]: Public doc probably launch at the same time, there's no direct alternative. Dev can use APNS for data message delivery and FireStore, firebase database, remote config for real time updates. But yeah we might be able to tuck in the public doc change for this.

File: Firebase/Messaging/CHANGELOG.md [hunk 2]
```diff
@@ -1,12 +1,13 @@
-# unreleased
+# 2020-02 -- v4.3.0
+- [changed] Deprecated Messaging Direct Channel and upstream API. Please use APNS channel instead for data message delivery. Use FireStore, real time database or other services for real time updates. The deprecated API will be removed in Firebase 7. (#4710)
```
Comment (egilmorez) [inline Firebase/Messaging/CHANGELOG.md]: We never really "productized" this feature with capital letters.  Suggest just "Deprecated FCM direct messaging via `shouldEstablishDirectChannel`. Instead, use APNs for downstream message delivery. "

Comment (egilmorez) [reply Firebase/Messaging/CHANGELOG.md]: Also suggest separating direct channel and upstream.  Would that work?  Then you'd have also:

- [changed] Deprecated upstream messaging API.  For realtime updates, use Cloud Firestore, Realtime Database, or other services.  The deprecated . . . "

**Pull request merged**